python/tank_vendor/shotgun_authentication/interactive_authentication.py

```diff
@@ -0,0 +1,432 @@
+# Copyright (c) 2015 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+UI and console based login for Toolkit.
+"""
+
+# Using "with" with the lock to make sure it is always released.
+
+from __future__ import with_statement
+from getpass import getpass
+import threading
+from .errors import AuthenticationError, AuthenticationDisabled, AuthenticationCancelled
+from . import session_cache
+
+
+# FIXME: Quick hack to easily disable logging in this module while keeping the
+# code compatible. We have to disable it by default because Maya will print all out
+# debug strings.
+if False:
+    # Configure logging
+    import logging
+    logger = logging.getLogger("sgtk.interactive_authentication")
+    logger.setLevel(logging.DEBUG)
+    logger.addHandler(logging.StreamHandler())
+else:
+    class logger:
+        @staticmethod
+        def debug(*args, **kwargs):
+            pass
+
+        @staticmethod
+        def info(*args, **kwargs):
+            pass
+
+        @staticmethod
+        def warning(*args, **kwargs):
+            pass
+
+        @staticmethod
+        def error(*args, **kwargs):
+            pass
+
+        @staticmethod
+        def exception(*args, **kwargs):
+            pass
+
+
+def _get_qt_state():
+    """
+    Returns the state of Qt: the librairies available and if we have a ui or not.
+    :returns: If Qt is available, a tuple of (QtCore, QtGui, has_ui_boolean_flag).
+              Otherwise, (None, None, None)
+    """
+    try:
+        from .ui.qt_abstraction import QtGui, QtCore
+    except ImportError:
+        return None, None, None
+    return QtCore, QtGui, QtGui.QApplication.instance() is not None
+
+
+def _create_invoker():
+    """
+    Create the object used to invoke function calls on the main thread when
+    called from a different thread.
+
+    :returns:  Invoker instance. If Qt is not available or there is no UI, no invoker will be returned.
+    """
+    QtCore, QtGui, has_ui = _get_qt_state()
+    # If we have a ui and we're not in the main thread, we'll need to send ui requests to the
+    # main thread.
+    if not QtCore or not QtGui or not has_ui:
```

Comment:
If there is no ui we can safely get user input from the console since those calls are thread safe (I think?)